core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -2380,31 +2382,83 @@ final class Stream[+F[_], +O] private[fs2] (private[fs2] val underlying: Pull[F,
     * scala> Stream(1, 2, 3, 4).sliding(2).toList
-    * res0: List[scala.collection.immutable.Queue[Int]] = List(Queue(1, 2), Queue(2, 3), Queue(3, 4))
+    * res0: List[fs2.Chunk[Int]] = List(Chunk(1, 2), Chunk(2, 3), Chunk(3, 4))
     * }}}
     * @throws scala.IllegalArgumentException if `n` <= 0
     */
-  def sliding(n: Int): Stream[F, collection.immutable.Queue[O]] = {
-    require(n > 0, "n must be > 0")
-    def go(
-        window: collection.immutable.Queue[O],
-        s: Stream[F, O]
-    ): Pull[F, collection.immutable.Queue[O], Unit] =
-      s.pull.uncons.flatMap {
-        case None => Pull.done
-        case Some((hd, tl)) =>
-          val (out, carry) = hd.scanLeftCarry(window)((w, i) => w.dequeue._2.enqueue(i))
-          Pull.output(out) >> go(carry, tl)
-      }
-    this.pull
-      .unconsN(n, true)
-      .flatMap {
-        case None => Pull.done
-        case Some((hd, tl)) =>
-          val window = hd.foldLeft(collection.immutable.Queue.empty[O])(_.enqueue(_))
-          Pull.output1(window) >> go(window, tl)
-      }
-      .stream
+  def sliding(n: Int): Stream[F, Chunk[O]] = sliding(n, 1)
+
+  /** Groups inputs in fixed size chunks by passing a "sliding window"
+    * of size with step over them. If the input contains less than or equal to
+    * `size` elements, only one chunk of this size will be emitted.
+    *
+    * @example {{{
+    * scala> Stream(1, 2, 3, 4, 5).sliding(2, 3).toList
+    * res0: List[fs2.Chunk[Int]] = List(Chunk(1, 2), Chunk(4, 5))
+    * scala> Stream(1, 2, 3, 4, 5).sliding(3, 2).toList
+    * res1: List[fs2.Chunk[Int]] = List(Chunk(1, 2, 3), Chunk(3, 4, 5))
+    * }}}
+    * @throws scala.IllegalArgumentException if `size` <= 0 | `step` <= 0
+    */
+  def sliding(size: Int, step: Int): Stream[F, Chunk[O]] = {
+    require(size > 0, "size must be > 0")
+    require(step > 0, "step must be > 0")
+
+    def stepNotSmallerThanSize(s: Stream[F, O], prev: Chunk[O]): Pull[F, Chunk[O], Unit] =
+      s.pull.uncons
+        .flatMap {
+          case None =>
+            if (prev.isEmpty) Pull.done
+            else Pull.output1(prev.take(size))
+          case Some((hd, tl)) =>
+            val buffer = ArrayBuffer.empty[Chunk[O]]
+            var (heads, tails) = (prev ++ hd).splitAt(step)
+            while (tails.nonEmpty) {
+              buffer += heads.take(size)
+              val (nHeads, nTails) = tails.splitAt(step)
+              heads = nHeads
+              tails = nTails
+            }
+            Pull.output(Chunk.buffer(buffer)) >> stepNotSmallerThanSize(tl, heads)
+        }
+
+    def stepSmallerThanSize(
+        s: Stream[F, O],
+        window: Chunk[O],
+        prev: Chunk[O]
+    ): Pull[F, Chunk[O], Unit] =
+      s.pull.uncons
+        .flatMap {
+          case None =>
+            if (prev.isEmpty) Pull.done
+            else Pull.output1((window ++ prev).take(size))
+          case Some((hd, tl)) =>
+            val buffer = ArrayBuffer.empty[Chunk[O]]
```

Comment:
I'm curious if here and above where `ArrayBuffer` is used if using `Chunk.Queue` would be better. Initialize buffer to an empty chunk.queue and append via `:+`

Reply:
I'm not sure, whether I’ve understood you correctly, but I’ve changed this and the part above to:
```
case Some((hd, tl)) =>
            var buffer = Chunk.Queue.empty[Chunk[O]]
            var w = window
            var (heads, tails) = (prev ++ hd).splitAt(step)
            while (tails.nonEmpty) {
              val wind = w ++ heads.take(step)
              buffer :+= Chunk.singleton(wind)
              w = wind.drop(step)
              heads = tails.take(step)
              tails = tails.drop(step)
            }

            Pull.output(buffer) >> stepSmallerThanSize(tl, w, heads)

```

I think, that Chunk.Queue’s job is to flatCombine Chunks effectively thus preserving chunk structure as close as possible. Here the task of buffer is different - extract distinct elements from input into a sequence of distinct elements(no flatten action)
Moreover, it has worser performance characteristics:
```
[info] StreamBenchmark.sliding     10  thrpt    5   650,337 ± 48,467  ops/s
[info] StreamBenchmark.sliding    100  thrpt    5   593,517 ± 26,828  ops/s
[info] StreamBenchmark.sliding   1000  thrpt    5   596,785 ± 24,417  ops/s
[info] StreamBenchmark.sliding  10000  thrpt    5  1370,548 ± 50,358  ops/s
```